mlrun/track/trackers/mlflow_tracker.py

```diff
@@ -0,0 +1,449 @@
+        :param is_offline: True if logging an offline run (importing), False if online run (tracking)
+        """
+        client = mlflow.MlflowClient()
+
+        # Create a temporary directory to log all data temporarily:
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # When running in offline mode (manually importing) we want to make sure there is an `artifact_path` set in
+            # the context, if not we will set it to this temporary directory so the user won't have garbage in his
+            # local work dir:
+            if is_offline and not context.artifact_path:
+                context.artifact_path = tmp_dir
+
+            # Set new MLRun related tags to the MLFlow run:
+            client.set_tag(
+                run_id=run.info.run_id, key="mlrun-context-id", value=context.uid
+            )
+            client.set_tag(
+                run_id=run.info.run_id, key="mlrun-project-name", value=context.project
+            )
+
+            # Get the MLFlow run's tags and save them as labels:
+            context.set_label(key="mlflow-user", value=run.data.tags.get("mlflow.user"))
+            context.set_label(
+                key="mlflow-run-name", value=run.data.tags.get("mlflow.runName")
+            )
+            context.set_label(key="mlflow-run-id", value=run.info.run_id)
+            context.set_label(key="mlflow-experiment-id", value=run.info.experiment_id)
+
+            # Get the MLFlow run's parameters and save as job parameters:
+            context._parameters.update(run.data.params)
+
+            # Get the MLFlow run's metrics and log them as results:
+            results = run.data.metrics
+            context.log_results(results=results)
+
+            # Import the MLFlow run's artifacts to MLRun (model are logged after the rest of artifacts
+            # so the artifacts can be registered as extra data in the models):
+            artifacts = {}
+            model_paths = []
+            for artifact in client.list_artifacts(run_id=run.info.run_id):
+                # Get the artifact's local path (MLFlow suggests that if the artifact is already in the local filesystem
+                # its local path will be returned:
+                artifact_local_path = mlflow.artifacts.download_artifacts(
+                    run_id=run.info.run_id,
+                    artifact_path=artifact.path,
+                )
+                # Check if the artifact is a model (will be logged after the artifacts):
+                if artifact.is_dir and os.path.exists(
+                    os.path.join(artifact_local_path, "MLmodel")
+                ):
+                    model_paths.append(artifact_local_path)
+                else:
+                    # Log the artifact:
+                    artifact = MLFlowTracker._log_artifact(
+                        context_or_project=context,
+                        key=pathlib.Path(artifact.path).name.replace(".", "_"),
```

Comment:
why do you normalize the key?

Reply:
they have the same name for files but with different extensions (png and json for example) so Ze'ev made the key to include the extension in the name.

Reply:
overwise i get errors later

Reply:
> they have the same name for files but with different extensions (png and json for example) so Ze'ev made the key to include the extension in the name.

aha. that worths a comment